MSC2314: A Method to Backfill Room State (#2314)
matrix-org/matrix-spec-proposals: proposals/2314-backfill-current-state.md

> # MSC-XXXX: Backfilling Current State

Reviewer: ```suggestion
# MSC2134: Backfilling Current State
```

(note: we don't use a dash, or space, or other delimiter here)

Author: thx

> If a server experiences data loss, it is difficult for it to recover membership in federation without having a user be reinvited, as backfill alone can not easily retrieve the data that is required to operate in a room (namely the current state and auth chain).

Reviewer 2: wrap at 80 chars please

Author: wrap... markdown?

Reviewer 2: yes please. It makes review easier to do because you can put comments near the text you're commenting on. It also makes the raw markdown easier to read.

Reviewer: @hawkowl bump

Author: done

> This MSC introduces S2S APIs to provide a given room's auth chain and current state events, provided the requesting server is in the room specified.

Reviewer 3: How do you see this API actually being used? The "current state" isn't really a concept that is used much in s2s, as if the server wants to start participating in the room it needs a set of most recent events and their state (from which the current state can be derived).

Author: Yes, it's asking another server for it's current state, in a situation where it knows it's joined to the room but is missing state (data loss).

Reviewer 3: But how does it actually use the current state? I don't think the current state is useful in rehydrating a room for example, as you need the extremities and the state at each of them.

Reviewer 2: hrm, it seems I missed this thread previously. @erikjohnston raises a good point - it is very hard to see how this API could be usefully used.

If we can't figure out a practical use for this change, we should instead reject it, and back out the changes in Synapse and Sytest.

@erikjohnston any further thoughts here?

Reviewer 3: It does feel like it *should* be useful, but I can't really see how. If this is mainly useful for rescuing a server that has suffered catastrophic memory lost then I think I'd like to see a step-by-step write up of the process of recovering from that, including any new APIs that need to be added.

Having a one stop "bootstrap this room for me" API might be useful, which returns the extremities, auth chain and state at each event (probably with some clever de-duplication there)

Reviewer 2: right then. In that case, I'm going to close this MSC, and back out the changes in sytest and synapse.

## Existing APIs

> If one knows the room ID and an event ID, `/_matrix/federation/v1/state/{roomId}?event_id={eventId}` can be used to retrieve the auth chain and current state events for that given event. However, this requires knowing an event ID (which cannot be assumed), as well as the version of the room (which can be assumed to be contained within the current state delivered). If an event arrives for a room that a Matrix server does not know about, the event ID could be used to backfill the state, but this is impractical for rooms which may be low-traffic yet valuable to the end-user.

Reviewer: ```suggestion
If one knows the room ID and an event ID, `/_matrix/federation/v1/state/{roomId}?event_id={eventId}` can be used to retrieve the auth chain and current state events for that given event. However, this requires knowing an event ID (which cannot be assumed), as well as the version of the room (which can be assumed to be contained within the current state delivered). If an event arrives for a room that a Matrix server does not know about, the event ID could be used to backfill the state, but this is impractical for rooms which may be low-traffic yet valuable to the end-user (such as DMs)
```

Author: fixed

## Proposal

Add a new v2 state API that returns the server's present auth chain and state PDUs if an event ID is not provided, as well as specifying the room version to ease the parsing of the given events.

```

> GET /_matrix/federation/v2/state/{roomId}

Reviewer 2: does this need to be v2? it looks backwards-compatible to me

Reviewer 2: cf https://github.com/matrix-org/matrix-doc/pull/2324/files#r340076892

{
  "room_version": "3",
  "auth_chain": [
    {
      "type": "m.room.minimal_pdu",
      "room_id": "!somewhere:example.org",
      "content": {
        "see_room_version_spec": "The event format changes depending on the room version."
      }
    }
  ],
  "pdus": [
    {
      "type": "m.room.minimal_pdu",
      "room_id": "!somewhere:example.org",
      "content": {
        "see_room_version_spec": "The event format changes depending on the room version."
      }
    }
  ]
}
```

> This requires the following changes to the existing API:

Reviewer: This section probably isn't needed - the explanations can be moved upwards into the paragraph before the example.

Author: done

- "room_version" is added to the response, defined as "Type: string. Optional, but required if event_id was not given. The version of the room that the state was queried of."

## Potential issues

Although the creating server is part of the room ID, a server using this API as a client may find that the target server does not presently know about the room (for example, it has been shut down or deleted). Finding servers that will successfully return results from this API is outside of the scope of this MSC.

Users may not know the room ID for a given room, only a room alias. Translating this alias into a room ID is outside of the scope of this MSC. Users of this API may want to use it in conjunction with `/_matrix/federation/v1/query/directory` to resolve aliases to room IDs as part of an end-user focused API.

Excessively large rooms may cause performance problems for servers implementing this API (including in its v1 incarnation). Discretionary rate limiting of this API may be required.

> ## Security considerations

Reviewer: servers might also have to worry about malicious attempts to get the server to join rooms. A server that really wants you to join a room could send the server an event over `/send`, therefore confusing the server into thinking it might be in the room. The victim server would then use this new API, find out some information about the room (possibly a falsified join event) and push the room through as joined. In theory at some point a signature check would fail, but another workaround could be to ask a couple servers in the room (if possible) to verify the state returned by a server.

# MSC2314: Backfilling Current State

If a server experiences data loss, it is difficult for it to recover membership
in federation without having a user be reinvited, as backfill alone can not
easily retrieve the data that is required to operate in a room (namely the
current state and auth chain).

This MSC introduces S2S APIs to provide a given room's auth chain and current
state events, provided the requesting server is in the room specified.

## Existing APIs

If one knows the room ID and an event ID,

> `/_matrix/federation/v1/state/{roomId}?event_id={eventId}` can be used to

Reviewer 2: ```suggestion
[`/_matrix/federation/v1/state/{roomId}?event_id={eventId}`](https://spec.matrix.org/v1.2/server-server-api/#get_matrixfederationv1stateroomid) can be used to
```

retrieve the auth chain and current state events for that given event. However,
this requires knowing an event ID (which cannot be assumed), as well as the
version of the room (which can be assumed to be contained within the current
state delivered). If an event arrives for a room that a Matrix server does not
know about, the event ID could be used to backfill the state, but this is
impractical for rooms which may be low-traffic yet valuable to the end-user
(such as direct messages).

## Proposal

> Make the `eventId` parameter to `/v1/state/{roomId}` optional, and return the

Reviewer 2: ```suggestion
Make the `eventId` parameter to [`/v1/state/{roomId}`](https://spec.matrix.org/v1.2/server-server-api/#get_matrixfederationv1stateroomid)
optional, and return the
```

room version to ease the parsing of the given events. If the `eventId` parameter
is not given, the receiving server is to instead use what it considers the

> room's current state. If `eventId` is not provided, `room_version` in the
> response becomes mandatory.

Reviewer 2: ```suggestion
room's current state. 

If `eventId` is not provided, `room_version` is added as a mandatory
response field, giving the version of the room.
```

```
GET /_matrix/federation/v1/state/{roomId}

{
  "room_version": "3",
  "auth_chain": [
    {
      "type": "m.room.minimal_pdu",
      "room_id": "!somewhere:example.org",
      "content": {
        "see_room_version_spec": "The event format changes depending on the room version."
      }
    }
  ],
  "pdus": [
    {
      "type": "m.room.minimal_pdu",
      "room_id": "!somewhere:example.org",
      "content": {
        "see_room_version_spec": "The event format changes depending on the room version."
      }
    }
  ]
}
```

## Potential issues

Although the creating server is part of the room ID, a server using this API as
a client may find that the target server does not presently know about the room
(for example, it has been shut down or deleted). Finding servers that will
successfully return results from this API is outside of the scope of this MSC.

Users may not know the room ID for a given room, only a room alias. Translating
this alias into a room ID is outside of the scope of this MSC. Users of this API

> may want to use it in conjunction with `/_matrix/federation/v1/query/directory`

Reviewer 2: ```suggestion
may want to use it in conjunction with [`/_matrix/federation/v1/query/directory`](https://spec.matrix.org/v1.2/server-server-api/#get_matrixfederationv1querydirectory)
```

to resolve aliases to room IDs as part of an end-user focused API.

Excessively large rooms may cause performance problems for servers implementing
this API (including in its v1 incarnation). Discretionary rate limiting of this
API may be required.

> ## Security considerations

Reviewer 2: possibly we need to add a note here about what happens if the `room_version` in the top level of the response differs from that in the `create` event in `pdus`:

```suggestion
## Security considerations

Requesting servers should be resilient to the `room_version` field in the response not
matching that in the `m.room.create` event in `pdus`.
```

state events, provided the requesting server is in the room specified.

## Existing APIs

If one knows the room ID and an event ID,
[`/_matrix/federation/v1/state/{roomId}?event_id={eventId}`](https://spec.matrix.org/v1.2/server-server-api/#get_matrixfederationv1stateroomid) can be used to
retrieve the auth chain and current state events for that given event. However,
this requires knowing an event ID (which cannot be assumed), as well as the
version of the room (which can be assumed to be contained within the current
state delivered). If an event arrives for a room that a Matrix server does not
know about, the event ID could be used to backfill the state, but this is
impractical for rooms which may be low-traffic yet valuable to the end-user
(such as direct messages).

## Proposal

Make the `eventId` parameter to [`/v1/state/{roomId}`](https://spec.matrix.org/v1.2/server-server-api/#get_matrixfederationv1stateroomid)
optional, and return the
room version to ease the parsing of the given events. If the `eventId` parameter
is not given, the receiving server is to instead use what it considers the
room's current state. 

> If `eventId` is not provided, `room_version` is added as a mandatory

Reviewer 4: Should we just go ahead and return the `room_version` even if the `eventId` query parameter is provided? It seems a bit confusing at first glance to only include it sometimes if you don't know the history of this MSC.

Reviewer 2: I wondered about that too. Synapse does in fact return `room_version` unconditionally as you suggest.

The question is: if I am writing a new implementation calling `/state/{roomId}?event_id={eventId}` am I allowed to assume `room_version` is returned? The two potential answers are:
 * no
 * only if the destination server implements a sufficiently recent version of the spec.

So the reason for this approach is an attempt to clarify the behaviour there. However, on thinking about it further, it seems to me that it really isn't much clearer at all, and I'll change this as you suggest.

Reviewer 2: done
